acme/open_spiel/agents/tf/dqn/learning.py

```diff
@@ -0,0 +1,211 @@
+      checkpoint: bool = True,
+  ):
+    """Initializes the learner.
+
+    Args:
+      network: the online Q network (the one being optimized)
+      target_network: the target Q critic (which lags behind the online net).
+      discount: discount to use for TD updates.
+      importance_sampling_exponent: power to which importance weights are raised
+        before normalizing.
+      learning_rate: learning rate for the q-network update.
+      target_update_period: number of learner steps to perform before updating
+        the target networks.
+      dataset: dataset to learn from, whether fixed or from a replay buffer (see
+        `acme.datasets.reverb.make_dataset` documentation).
+      huber_loss_parameter: Quadratic-linear boundary for Huber loss.
+      replay_client: client to replay to allow for updating priorities.
+      counter: Counter object for (potentially distributed) counting.
+      logger: Logger object for writing logs to.
+      checkpoint: boolean indicating whether to checkpoint the learner.
+    """
+
+    # Internalise agent components (replay buffer, networks, optimizer).
+    # TODO(b/155086959): Fix type stubs and remove.
+    self._iterator = iter(dataset)  # pytype: disable=wrong-arg-types
+    self._network = network
+    self._target_network = target_network
+    self._optimizer = snt.optimizers.Adam(learning_rate)
+    self._replay_client = replay_client
+
+    # Internalise the hyperparameters.
+    self._discount = discount
+    self._target_update_period = target_update_period
+    self._importance_sampling_exponent = importance_sampling_exponent
+    self._huber_loss_parameter = huber_loss_parameter
+
+    # Learner state.
+    self._variables: List[List[tf.Tensor]] = [network.trainable_variables]
+    self._num_steps = tf.Variable(0, dtype=tf.int32)
+
+    # Internalise logging/counting objects.
+    self._counter = counter or counting.Counter()
+    self._logger = logger or loggers.TerminalLogger('learner', time_delta=1.)
+
+    # Create a snapshotter object.
+    if checkpoint:
+      self._snapshotter = tf2_savers.Snapshotter(
+          objects_to_save={'network': network}, time_delta_minutes=60.)
+    else:
+      self._snapshotter = None
+
+    # Do not record timestamps until after the first learning step is done.
+    # This is to avoid including the time it takes for actors to come online and
+    # fill the replay buffer.
+    self._timestamp = None
+
+  @tf.function
+  def _step(self) -> Dict[str, tf.Tensor]:
+    """Do a step of SGD and update the priorities."""
+
+    # Pull out the data needed for updates/priorities.
+    inputs = next(self._iterator)
+    o_tm1, a_tm1, r_t, d_t, o_t, e_t = inputs.data
+    keys, probs = inputs.info[:2]
+
+    legal_actions_t = e_t.legal_actions
+    terminals_t = e_t.terminals
+
+    with tf.GradientTape() as tape:
+      # Evaluate our networks.
+      q_tm1 = self._network(o_tm1)
+      q_t_value = self._target_network(o_t)
```

Comment:
Is this the main change for DQN in this case? IE I wonder (as I think @fastturtle was alluding to) if the `legal_actions` mask is moved into the observations directly if this logic can then be moved into the policy (perhaps via a wrapper around a "raw q-function network"). This would then allow a broader array of potential algorithms to be applied in this setting.